traceroute-eval/scripts/parse-traceroutes.py

```diff
@@ -0,0 +1,199 @@
+
+    # Each line represents one hop
+    for hop in lines[2:-1]:
+        tokens = hop.split()
+        ttl = tokens[0]
+        ips = []
+        rtts = []
+        
+        # A token is either an IP, latency, "ms", or "*"
+        for token in tokens[1:]:
+            if token == "*":
+                ips.append("*")
+                rtts.append("*")
+            elif ip_pattern.match(token):
+                ips.append(token)
+            elif ping_pattern.match(token):
+                rtts.append(token)
+                # If same IP reached
+                if len(rtts) > len(ips):
+                    # Find last non-"*" IP
+                    last_ip = [ip for ip in ips if ip != "*"][-1]
+                    ips.append(last_ip)
+        
+        # Add hop to data dictionary
+        json_dict[HOPS].append({
+            TTL: ttl, IP: ips, RTT: rtts
+        })
+    
+    return json_dict
+
+def tcp_parser(traceroute):
+    """
+    Parses one TCP traceroute.
+
+    Args: 
+        traceroute (str): the full traceroute output
+    Returns:
+        json_dict (dict): dict containing relevant information
+    """
+    return base_parser(traceroute, "tcp")
+
+def udp_parser(traceroute):
+    """
+    Parses one UDP traceroute.
+
+    Args: 
+        traceroute (str): the full traceroute output
+    Returns:
+        json_dict (dict): dict containing relevant information
+    """
+    return base_parser(traceroute, "udp")
+
+def icmp_parser(traceroute):
+    """
+    Parses one ICMP traceroute.
+
+    Args: 
+        traceroute (str): the full traceroute output
+    Returns:
+        json_dict (dict): dict containing relevant information
+    """
+    return base_parser(traceroute, "icmp")
+
+def paris_parser(traceroute):
+    """
+    Parses one ICMP traceroute.
+
+    Args: 
+        traceroute (str): the full traceroute output
+    Returns:
+        json_dict (dict): dict containing relevant information
+    """
+    return base_parser(traceroute, "paris")
+
+def dublin_parser(traceroute):
+    """
+    Parses one ICMP traceroute.
+
+    Args: 
+        traceroute (str): the full traceroute json as a string
+    Returns:
+        json_dict (dict): dict containing relevant information
+    """
+    flow_dicts = []
+    traceroute = json.loads(traceroute)
+
+    # JSON is a dict with flow_id keys
+    for flow_id, flow in traceroute["flows"].items():
+        json_dict = {}
+        json_dict[TS] = flow[0]["sent"]["timestamp"]
+        json_dict[DEST] = flow[0]["sent"]["ip"]["dst"]
+        json_dict[TYPE] = "dublin"
+        json_dict["FlowID"] = flow_id
+        json_dict[HOPS] = []
+        
+        # The value of the flow_id is a list of hop dictionaries
+        for hop in flow:
+            ttl = hop["sent"]["ip"]["ttl"]
+            try:
+                rtt = hop["rtt_usec"]/1000
+                ip = hop["received"]["ip"]["src"]
+            except TypeError:
+                rtt = "*"
+                ip = "*"
```

Comment:
Instead of TypeError -- check for `null` looks like thats the default error we get from dublin in the trace.json (correct me if I'm wrong)
This is a snippet from a trace.json where it failed:
```
                "33453" :
                [
                        {
                                "flowhash" : 34949,
                                "is_last" : true,
                                "name" : "",
                                "received" : null,
                                "rtt_usec" : null,
                                "sent" :
                                {
                                        "ip" :
                                        {
                                                "dst" : "1.1.1.1",
                                                "src" : "141.212.123.234",
                                                "ttl" : 1
                                        },
                                        "timestamp" : "1678725626.461646",
                                        "udp" :
                                        {
                                                "dport" : 33453,
                                                "sport" : 12345
                                        }
                                }
                        }
                ]
```